File: SVI/Code/SVI_SUR/R/SVI_stochasticVolatilityNatGrad.R
library(mvtnorm)    # for multivariate Gaussian distribution
library(matrixcalc) # for duplication matrix and elimination matrix
library(MCMCpack)   # for vech and xpnd
library(Matrix)     # for block-diagonal operator
logh <- function(n,alpha,b,y,lam,phi,psi,sig.a2,sig.l2,sig.p2) {
  v <- -0.5*n*lam-0.5*exp(alpha)*sum(b)-0.5*sum(y*y*exp(-lam-exp(alpha)*b))-
       0.5*sum((b[2:length(b)]-b[1:(length(b)-1)])*(b[2:length(b)]-b[1:(length(b)-1)]))-0.5*log(1-phi*phi)-
       0.5*(1-phi*phi)*b[1]*b[1]-alpha*alpha/(2*sig.a2)-lam*lam/(2*sig.l2)-psi*psi/(2*sig.p2)
  v     
}

n <- 200
d <- n+3
phi_true <- 0.3
lam_true <- 1.24
sig_true <- -1.73
b1_true <- rnorm(1,0,sqrt(1/(1-phi_true*phi_true)))
b_true <- rep(0,n)
b_true[1] <- b1_true
y <- rep(0,n)
y[1] <- rnorm(1,0,exp(0.5*(lam_true+sig_true*b_true[1])))
for (t in 2:n) {
  b_true[t] <- rnorm(1,phi_true*b_true[t-1],1)
  y[t] <- rnorm(1,0,exp(0.5*(lam_true+sig_true*b_true[t])))
}


# initialize hyperparameters
sig.a2 <- 1
sig.l2 <- 1
sig.p2 <- 1


# initialize variational parameters
mu.q  <- rep(0,n+3)
sig.q <- crossprod(matrix(runif(d*d),d,d))
invsig.q <- chol2inv(chol(sig.q))

D_mat <- duplication.matrix(d)
D_plus <- ginv(D)
count <- 0
lb <- 0
lbc <- c()
S <- 40
nIter <- 4000

gauss.lambda1 <- c(solve(sig.q,mu.q))
guass.lambda2 <- -0.5*crossprod(D,as.vector(invsig.q))
gauss.lambda <- c(gauss.lambda1,gauss.lambda2)


mu.q.storage <- matrix(0,d,nIter)
count <- 0
# start inference
for (i in 1:nIter) {
  count <- count+1
  cat(count,'th iteration\n')
  gauss.lambda.old <- gauss.lambda
  mu.q.old <- mu.q
  sig.q.old <- sig.q
  rho <- 1/(5+i)
  M_mat <- 2*D_plus%*%(mu.q%x%diag(d))
  S_mat <- 2*D_plus%*%tcrossprod((sig.q%x%sig.q),D_plus)
  aux   <- solve(S_mat,M_mat)
  gauss.invfisherinformation <- rbind(cbind(invsig.q+crossprod(M_mat,aux),-t(aux)),cbind(-aux,solve(S_mat)))

  theta.sample <- rmvnorm(S,mean=mu.q,sigma=sig.q)
  gauss_m      <- rowMeans(apply(theta.sample,1,function(theta) c(theta-mu.q,vech(tcrossprod(theta)-sig.q-tcrossprod(mu.q)))))
  gauss.grad   <- 0
  lb           <- 0
  for (j in 1:S) {
    theta  <- theta.sample[j,]
    b      <- theta[1:n]
    alpha  <- theta[n+1]
    lam    <- theta[n+2]
    psi    <- theta[n+3]
    phi    <- exp(psi)/(1+exp(psi))
    sig    <- exp(alpha)
    logh.v <- logh(n,alpha,b,y,lam,phi,psi,sig.a2,sig.l2,sig.p2)
    logq.v <- dmvnorm(theta,mean=mu.q,sigma=sig.q,log=TRUE)
    f.v    <- logh.v-logq.v
    lb     <- lb+f.v
    gauss.grad <- gauss.grad+(c(theta-mu.q,vech(tcrossprod(theta)-sig.q-tcrossprod(mu.q)))-gauss.m)*f.v
  }
  gauss.grad <- gauss.grad/(S-1)
  gauss.lambda <- (1-rho)*gauss.lambda+rho*c(gauss.invfisherinformation%*%gauss.grad)
  lambda1 <- gauss.lambda[1:d]
  lambda2 <- gauss.lambda[(d+1):length(gauss.lambda)]
  sig.q   <- -0.5*solve(matrix(c(crossprod(D_plus,lambda2)),d,d,byrow=FALSE))
  mu.q    <- c(sig.q%*%lambda1)
  if (any(eigen(sig.q)$values<0)) {
    gauss.lambda <- gauss.lambda.old
    sig.q        <- sig.q.old
    mu.q         <- mu.q.old
  } else {
    invsig.q     <- chol2inv(chol(sig.q))
  }
  lb <- lb/S
  lbc <- c(lbc,lb)
  mu.q.storage[,i] <- mu.q
  cat('lb: ',lb,'\n')
}

File: SVI/Code/SVI_SUR/MATLAB/SVI_stochasticVolatilityNatGrad.m
n = 200;
phi_true = 0.3;
lam_true = 1.24;
sig_true = -1.73;
b1_true = randn(1,1)*sqrt(1/(1-phi_true*phi_true));
b_true = zeros(n,1);
b_true(1) = b1_true;
y = zeros(n,1);
y(1) = randn(1,1)*exp(0.5*(lam_true+sig_true*b_true(1)));
for t = 2:n
   b_true(t) = randn(1,1)+phi_true*b_true(t-1);
   y(t) = randn(1,1)*exp(0.5*(lam_true+sig_true*b_true(t)));
end

sig_a2 = 1;
sig_l2 = 1;
sig_p2 = 1;

mu_q  = zeros(n+3,1);
sig_q = eye(n+3);
D_mat = duplication_matrix_full(n+3);
D_plus = duplication_matrix_MP_inv(n+3);
invsig_q = invpd(sig_q);
gauss_lambda1 = sig_q\mu_q;
gauss_lambda2 = -0.5*D_mat'*invsig_q(:);
gauss_lambda = [gauss_lambda1;gauss_lambda2];
% T_mat = chol(invsig_q,'lower');
% T_mat_transpose = T_mat';
% T_mat_prime = T_mat;
% for k = 1:(n+3)
%    T_mat_prime(k,k) = log(T_mat(k,k));
% end
niter = 10000;
S = 200; % number of simulations
count = 0;
lbc = [];
mu_q_storage = zeros(n+3,niter);

% start inference
for i = 1:niter
    gauss_lambda_old = gauss_lambda;
    mu_q_old = mu_q;
    sig_q_old = sig_q;
    rho = 1/(5+i);
    M_mat = 2*D_plus*kron(mu_q,eye(n+3));
    S_mat = 2*D_plus*kron(sig_q,sig_q)*D_plus';
    aux = S_mat\M_mat;
    gauss_invfisherinformation = [invsig_q+M_mat'*aux -aux';-aux eye(size(S_mat))/S_mat];
    theta = mvnrnd(mu_q,sig_q,S);
   
end